blobfuse/utilities.cpp

```diff
@@ -481,9 +494,13 @@ int azs_access(const char * /*path*/, int /*mask*/)
     return 0; // permit all access
 }
 
-int azs_fsync(const char * /*path*/, int /*isdatasync*/, struct fuse_file_info * /*fi*/)
+int azs_fsync(const char * path, int /*isdatasync*/, struct fuse_file_info *fi)
 {
-    return 0; // Skip for now
```

Comment:
we need this for directories also

Reply:
I tried to test 'sync' command on directory but did not get any call in blobfuse for the same. 'sync <filename>' consistently resulted in fsync call on our end.

Reply:
Also as you can see this fsync is coming with a handle info, which was returned back in 'open' file call so not sure if libfuse supports fsync of directory or not. We do not support opendir call in blobfuse.

Reply:
The AML CI Jupyter team will use the Jupyter Notebook command to issue a “sync <filename>” whenever they want to ensure file contents are not stale. They will issue a  “sync dirname” command to ensure git status results are not stale, in  other ways refresh dir listing.

These will invoke the blobfuse fsync  or fsyncdir respectively.

Refer to Workitem: https://msazure.visualstudio.com/One/_backlogs/backlog/XStore-DevExp/Epics/?workitem=10196695
The fsync will invalidate file attribute cache on the kernel and the memory and delete the content cache for the file from the disk.

The fsyncdir will invalidate the directory list of attributes cache for all its files and subdirs.\

Reply:
As of now blobfuse does not register and handel fsyncdir, this will need new registeration then. Also handling fsyncdir also means we need to handle opendir and closedir, which do not have any implementation as of now.

Reply:
Once fsyncdir is registered, we are getting the dir name as well in the api so relying on that we can ignore opendir and closedir. As per the task item we just need to invalidate the attr-cache on fsyncdir so have done code changes for that. However there is one concern, even if we invalidate the attr-cache, getattr on file may still rely on disk cached file for attributes so we need to do sync on those cached files seperatly.

Reply:
without fysndir how will the dir attributes be cleared even if you add the code to AttrCacheBfsCllient?
Also when you delete the attribute cache for a dir, you have to delete the attribute for all the files and subdirs in it too. The best thing at this time is to remove remove the whole attribute invalidation from  the PR and keep it to streaming alone.

Reply:
In fsyncdir we are getting the directory path being synced. Using that in attr-cache layer we are iterating through all elements having that path and invalidating the cache. So its working as per the expectation given in the task. In task it only talks about invalidating the attr-cache and not deleting files on that path to be removed from the disk cache.